ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HeaderStateHistory.hs

```diff
@@ -107,12 +124,81 @@ cast (HeaderStateHistory history) =
 rewind ::
      forall blk. (HasAnnTip blk)
   => Point blk
-  -> HeaderStateHistory blk -> Maybe (HeaderStateHistory blk)
-rewind p (HeaderStateHistory history) = HeaderStateHistory <$>
-    AS.rollback
+  -> HeaderStateHistory blk
+  -> Maybe (HeaderStateHistory blk, Maybe (HeaderStateWithTime blk))
+rewind p (HeaderStateHistory history) = do
+    (prefix, suffix) <- AS.splitAfterMeasure
       (pointSlot p)
-      ((== p) . either headerStatePoint headerStatePoint)
+      ((== p) . join either (headerStatePoint . hswtHeaderState))
```

Comment:
Does `(headerStatePoint . hswtHeaderState) . either id id` work? That might be easier for many readers.